File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/nutrition_record_mappers.dart
import 'package:health_connector_core/health_connector_core.dart'
    show
        Energy,
        HealthRecord,
        HealthRecordId,
        Mass,
        NutritionRecord,
        sinceV1_1_0;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/meal_type_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show EnergyDto, HealthDataTypeDto, MassDto, NutritionRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [NutritionRecord] to [NutritionRecordDto].
@sinceV1_1_0
@internal
extension NutritionRecordToDto on NutritionRecord {
  NutritionRecordDto toDto() {
    return NutritionRecordDto(
      id: id.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      healthDataType: HealthDataTypeDto.nutrition,
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDto(),
      foodName: foodName,
      mealType: mealType.toDto(),
      energy: energy?.toDto() as EnergyDto?,
      protein: protein?.toDto() as MassDto?,
      totalCarbohydrate: totalCarbohydrate?.toDto() as MassDto?,
      totalFat: totalFat?.toDto() as MassDto?,
      saturatedFat: saturatedFat?.toDto() as MassDto?,
      monounsaturatedFat: monounsaturatedFat?.toDto() as MassDto?,
      polyunsaturatedFat: polyunsaturatedFat?.toDto() as MassDto?,
      cholesterol: cholesterol?.toDto() as MassDto?,
      dietaryFiber: dietaryFiber?.toDto() as MassDto?,
      sugar: sugar?.toDto() as MassDto?,
      vitaminA: vitaminA?.toDto() as MassDto?,
      vitaminB6: vitaminB6?.toDto() as MassDto?,
      vitaminB12: vitaminB12?.toDto() as MassDto?,
      vitaminC: vitaminC?.toDto() as MassDto?,
      vitaminD: vitaminD?.toDto() as MassDto?,
      vitaminE: vitaminE?.toDto() as MassDto?,
      vitaminK: vitaminK?.toDto() as MassDto?,
      thiamin: thiamin?.toDto() as MassDto?,
      riboflavin: riboflavin?.toDto() as MassDto?,
      niacin: niacin?.toDto() as MassDto?,
      folate: folate?.toDto() as MassDto?,
      biotin: biotin?.toDto() as MassDto?,
      pantothenicAcid: pantothenicAcid?.toDto() as MassDto?,
      calcium: calcium?.toDto() as MassDto?,
      iron: iron?.toDto() as MassDto?,
      magnesium: magnesium?.toDto() as MassDto?,
      manganese: manganese?.toDto() as MassDto?,
      phosphorus: phosphorus?.toDto() as MassDto?,
      potassium: potassium?.toDto() as MassDto?,
      selenium: selenium?.toDto() as MassDto?,
      sodium: sodium?.toDto() as MassDto?,
      zinc: zinc?.toDto() as MassDto?,
      caffeine: caffeine?.toDto() as MassDto?,
    );
  }
}

/// Converts [NutritionRecordDto] to domain nutrition records.
@sinceV1_1_0
@internal
extension NutritionRecordDtoToDomain on NutritionRecordDto {
  HealthRecord toDomain() {
    final id = this.id?.toDomain() ?? HealthRecordId.none;
    final time = DateTime.fromMillisecondsSinceEpoch(startTime);
    final zoneOffsetSeconds = startZoneOffsetSeconds;
    final metadata = this.metadata.toDomain();
    final foodName = this.foodName;
    final mealType = this.mealType.toDomain();

    switch (healthDataType) {
      case HealthDataTypeDto.nutrition:
        return NutritionRecord(
          id: id,
          startTime: time,
          endTime: DateTime.fromMillisecondsSinceEpoch(endTime),
          startZoneOffsetSeconds: zoneOffsetSeconds,
          endZoneOffsetSeconds: endZoneOffsetSeconds,
          metadata: metadata,
          foodName: foodName,
          mealType: mealType,
          energy: energy?.toDomain() as Energy?,
          protein: protein?.toDomain() as Mass?,
          totalCarbohydrate: totalCarbohydrate?.toDomain() as Mass?,
          totalFat: totalFat?.toDomain() as Mass?,
          saturatedFat: saturatedFat?.toDomain() as Mass?,
          monounsaturatedFat: monounsaturatedFat?.toDomain() as Mass?,
          polyunsaturatedFat: polyunsaturatedFat?.toDomain() as Mass?,
          cholesterol: cholesterol?.toDomain() as Mass?,
          dietaryFiber: dietaryFiber?.toDomain() as Mass?,
          sugar: sugar?.toDomain() as Mass?,
          vitaminA: vitaminA?.toDomain() as Mass?,
          vitaminB6: vitaminB6?.toDomain() as Mass?,
          vitaminB12: vitaminB12?.toDomain() as Mass?,
          vitaminC: vitaminC?.toDomain() as Mass?,
          vitaminD: vitaminD?.toDomain() as Mass?,
          vitaminE: vitaminE?.toDomain() as Mass?,
          vitaminK: vitaminK?.toDomain() as Mass?,
          thiamin: thiamin?.toDomain() as Mass?,
          riboflavin: riboflavin?.toDomain() as Mass?,
          niacin: niacin?.toDomain() as Mass?,
          folate: folate?.toDomain() as Mass?,
          biotin: biotin?.toDomain() as Mass?,
          pantothenicAcid: pantothenicAcid?.toDomain() as Mass?,
          calcium: calcium?.toDomain() as Mass?,
          iron: iron?.toDomain() as Mass?,
          magnesium: magnesium?.toDomain() as Mass?,
          manganese: manganese?.toDomain() as Mass?,
          phosphorus: phosphorus?.toDomain() as Mass?,
          potassium: potassium?.toDomain() as Mass?,
          selenium: selenium?.toDomain() as Mass?,
          sodium: sodium?.toDomain() as Mass?,
          zinc: zinc?.toDomain() as Mass?,
          caffeine: caffeine?.toDomain() as Mass?,
        );
      default:
        throw ArgumentError('Invalid DTO for $NutritionRecordDto: $this');
    }
  }
}

/// Maps individual nutrient domain record to [NutritionRecordDto].

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/NutritionRecordMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.NutritionRecord
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.HealthDataTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.NutritionRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a [NutritionRecordDto] to a Health Connect [NutritionRecord] object.
 */
internal fun NutritionRecordDto.toHealthConnect(): NutritionRecord {
    val startTimeInstant = Instant.ofEpochMilli(startTime)
    val endTimeInstant = Instant.ofEpochMilli(endTime)
    val startZoneOffset =
        startZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) } ?: ZoneOffset.UTC
    val endZoneOffset = endZoneOffsetSeconds?.let {
        ZoneOffset.ofTotalSeconds(it.toInt())
    } ?: ZoneOffset.UTC

    // Build NutritionRecord based on healthDataType
    return when (healthDataType) {
        HealthDataTypeDto.NUTRITION -> NutritionRecord(
            startTime = startTimeInstant,
            endTime = endTimeInstant,
            startZoneOffset = startZoneOffset,
            endZoneOffset = endZoneOffset,
            name = foodName,
            mealType = mealType.toHealthConnect(),
            energy = energy?.toHealthConnect(),
            protein = protein?.toHealthConnect(),
            totalCarbohydrate = totalCarbohydrate?.toHealthConnect(),
            totalFat = totalFat?.toHealthConnect(),
            saturatedFat = saturatedFat?.toHealthConnect(),
            monounsaturatedFat = monounsaturatedFat?.toHealthConnect(),
            polyunsaturatedFat = polyunsaturatedFat?.toHealthConnect(),
            cholesterol = cholesterol?.toHealthConnect(),
            dietaryFiber = dietaryFiber?.toHealthConnect(),
            sugar = sugar?.toHealthConnect(),
            vitaminA = vitaminA?.toHealthConnect(),
            vitaminB6 = vitaminB6?.toHealthConnect(),
            vitaminB12 = vitaminB12?.toHealthConnect(),
            vitaminC = vitaminC?.toHealthConnect(),
            vitaminD = vitaminD?.toHealthConnect(),
            vitaminE = vitaminE?.toHealthConnect(),
            vitaminK = vitaminK?.toHealthConnect(),
            thiamin = thiamin?.toHealthConnect(),
            riboflavin = riboflavin?.toHealthConnect(),
            niacin = niacin?.toHealthConnect(),
            folate = folate?.toHealthConnect(),
            biotin = biotin?.toHealthConnect(),
            pantothenicAcid = pantothenicAcid?.toHealthConnect(),
            calcium = calcium?.toHealthConnect(),
            iron = iron?.toHealthConnect(),
            magnesium = magnesium?.toHealthConnect(),
            manganese = manganese?.toHealthConnect(),
            phosphorus = phosphorus?.toHealthConnect(),
            potassium = potassium?.toHealthConnect(),
            selenium = selenium?.toHealthConnect(),
            sodium = sodium?.toHealthConnect(),
            zinc = zinc?.toHealthConnect(),
            caffeine = caffeine?.toHealthConnect(),
            metadata = metadata.toHealthConnect(id),
        )

        else -> throw IllegalArgumentException(
            "Unsupported healthDataType for NutritionRecordDto: $healthDataType",
        )
    }
}

/**
 * Converts NutritionRecord to a unified NutritionRecordDto for combined nutrition.
 *
 * Creates a NutritionRecordDto with healthDataType = NUTRITION and all non-null fields populated.
 */
internal fun NutritionRecord.toDto(): NutritionRecordDto = NutritionRecordDto(
    id = metadata.id,
    metadata = metadata.toDto(),
    startTime = startTime.toEpochMilli(),
    endTime = endTime.toEpochMilli(),
    healthDataType = HealthDataTypeDto.NUTRITION,
    startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
    endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
    foodName = name,
    mealType = mealType.toMealTypeDto(),
    energy = energy?.toDto(),
    protein = protein?.toDto(),
    totalCarbohydrate = totalCarbohydrate?.toDto(),
    totalFat = totalFat?.toDto(),
    saturatedFat = saturatedFat?.toDto(),
    monounsaturatedFat = monounsaturatedFat?.toDto(),
    polyunsaturatedFat = polyunsaturatedFat?.toDto(),
    cholesterol = cholesterol?.toDto(),
    dietaryFiber = dietaryFiber?.toDto(),
    sugar = sugar?.toDto(),
    vitaminA = vitaminA?.toDto(),
    vitaminB6 = vitaminB6?.toDto(),
    vitaminB12 = vitaminB12?.toDto(),
    vitaminC = vitaminC?.toDto(),
    vitaminD = vitaminD?.toDto(),
    vitaminE = vitaminE?.toDto(),
    vitaminK = vitaminK?.toDto(),
    thiamin = thiamin?.toDto(),
    riboflavin = riboflavin?.toDto(),
    niacin = niacin?.toDto(),
    folate = folate?.toDto(),
    biotin = biotin?.toDto(),
    pantothenicAcid = pantothenicAcid?.toDto(),
    calcium = calcium?.toDto(),
    iron = iron?.toDto(),
    magnesium = magnesium?.toDto(),
    manganese = manganese?.toDto(),
    phosphorus = phosphorus?.toDto(),
    potassium = potassium?.toDto(),
    selenium = selenium?.toDto(),
    sodium = sodium?.toDto(),
    zinc = zinc?.toDto(),
    caffeine = caffeine?.toDto(),
)
